projects/solar/test/test_fan_control_solar.c

```diff
@@ -0,0 +1,342 @@
+    data_store_set(DATA_POINT_TEMPERATURE(i), ABOVE_TEMPERATURE_THRESHOLD_DC);
+  }
+  fan_control_process_event(&s_data_ready);
+
+  // Test that the full_speed pin is indeed active low
+  gpio_get_state(&s_full_speed_addr, &s_gpio_state);
+  TEST_ASSERT_EQUAL(GPIO_STATE_LOW, s_gpio_state);
+
+  // Set temperatures below overtemp threshold
+  for (Mppt i = 0; i < 6; i++) {
+    data_store_set(DATA_POINT_TEMPERATURE(i), BELOW_TEMPERATURE_THRESHOLD_DC);
+  }
+  fan_control_process_event(&s_data_ready);
+
+  // Test that full_speed pin is inactive (high)
+  gpio_get_state(&s_full_speed_addr, &s_gpio_state);
+  TEST_ASSERT_EQUAL(GPIO_STATE_HIGH, s_gpio_state);
+}
+
+// Test that full_speed set inactive (high)
+// when MPPTs no longer detecting overtemp status
+void test_fan_overtemp_fault_fullspeed_handling_case_nolongerovtempstatus(void) {
+  FanControlSolarSettings settings = {
+    .overtemp_addr = s_overtemp_addr,
+    .fan_fail_addr = s_fan_fail_addr,
+    .full_speed_addr = s_full_speed_addr,
+    .full_speed_temp_threshold_dC = s_full_speed_temp_threshold,
+    .mppt_count = s_mppt_count,
+  };
+  TEST_ASSERT_OK(fan_control_init(&settings));
+
+  // Initialize all mppt statuses to nominal
+  for (int i = 0; i < 6; i++) {
+    data_store_set(DATA_POINT_MPPT_STATUS(i), SPV1020_NOMINAL_MASK);
+  }
+
+  // First trigger full_speed pin to go active
+  data_store_set(DATA_POINT_MPPT_STATUS(5), SPV1020_OVT_MASK);
+  fan_control_process_event(&s_data_ready);
+
+  // Test that the full_speed pin is indeed active low
+  gpio_get_state(&s_full_speed_addr, &s_gpio_state);
+  TEST_ASSERT_EQUAL(GPIO_STATE_LOW, s_gpio_state);
+
+  // Set status to nominal
+  data_store_set(DATA_POINT_MPPT_STATUS(5), SPV1020_NOMINAL_MASK);
+  fan_control_process_event(&s_data_ready);
+
+  // Test that full_speed pin is inactive (high)
+  gpio_get_state(&s_full_speed_addr, &s_gpio_state);
+  TEST_ASSERT_EQUAL(GPIO_STATE_HIGH, s_gpio_state);
+}
+
+// Test that initializing with invalid settings fails gracefully.
+void test_invalid_settings(void) {
+  // NULL init test
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS, fan_control_init(NULL));
+
+  // too many MPPTs test
+  GpioAddress overtemp_addr = { .port = GPIO_PORT_A, .pin = 0 };
+  GpioAddress fan_fail_addr = { .port = GPIO_PORT_A, .pin = 1 };
+  GpioAddress full_speed_addr = { .port = GPIO_PORT_A, .pin = 2 };
+
+  FanControlSolarSettings invalid_settings = {
+    .overtemp_addr = overtemp_addr,
+    .fan_fail_addr = fan_fail_addr,
+    .full_speed_addr = full_speed_addr,
+    .full_speed_temp_threshold_dC = 4,
+    .mppt_count = MAX_SOLAR_BOARD_MPPTS + 1,
+  };
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS, fan_control_init(&invalid_settings));
+}
+
+// Test that processing a non DATA_READY_EVENT returns false (fails gracefully)
+void test_not_data_ready_event(void) {
+  // NULL passed in
+  TEST_ASSERT_FALSE(fan_control_process_event(NULL));
+
+  // Non DATA_READY_EVENT passed in
+  Event e;
```

Comment:
Could you initialize this to something? Like `Event e = { DATA_READY_EVENT + 1, 0 }`. Otherwise `e` has a value of uninitialized garbage, which could in theory occasionally make `e.id` by chance get initialized to `DATA_READY_EVENT`'s value.